workers/scanner.go

```diff
@@ -238,7 +245,27 @@ func (s *Scanner) threadedScanAndPin() {
 		s.staticLogger.Tracef("Start scanning")
 		s.managedRefreshDryRun()
 		s.managedRefreshMinPinners()
-		s.managedPinUnderpinnedSkylinks()
+		s.managedResetSkippedSkylinks()
+		s.managedResetStats()
+
+		// Start a thread that will print intermediate scanning statistics.
+		statsCh := make(chan struct{})
+		go s.threadedPrintStats(statsCh)
+
+		// Start N threads that will scan for underpinned skylinks and repin
+		// them. It's possible that at first all of those start pinning skylinks
+		// without properly respecting the MaxRepairingSkylinks limit. That's
+		// expected and chosen because of the simplicity of the implementation.
+		var wg sync.WaitGroup
+		for i := 0; i < s.staticScannerThreads; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				s.managedPinUnderpinnedSkylinks()
+			}()
+		}
+		wg.Wait()
+		close(statsCh)
```

Comment:
you might want to wait for the stats thread to be done as well before returning.

Goroutines which run in the background and write to files can panic if the main shutdown thread closes the file and the background thread wants to access the filehandle after that.

Reply:
Good point, I'll add it to the threadgroup counter. It seems like all threaded methods should start with `defer <tg>.done()`.